HPCC-18038 Remove the version option from the publish and java command

mayx opened the pull request:
- Remove the --version option from the usage message
- Don't pass in any version info when loading dependencies,
  pass in 0 instead.
- Don't validate version command option
- Replace --version with --build for printing out build info
- Remove the --version option from the batch file for ECL IDE

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@rpastrana @afishbeck Please review when you have a chance.

rpastrana commented on tools/esdlcmd/esdlcmd_shell.cpp:
```diff
@@ -180,7 +180,7 @@ bool EsdlCMDShell::parseCommandLineOptions(ArgvIterator &iter)
 void EsdlCMDShell::usage()
 {
     fprintf(stdout,"\nUsage:\n"
-        "    esdl [--version] <command> [<args>]\n\n"
```
We shouldn't remove the esdl --version command (this prints out the version of the tool)

rpastrana replied:
We should provide usage for the new "build" command, but also consider keeping this version command to remain consistent across various HPCC command line tools (ecl, eclcc, etc).

mayx replied:
I did add usage for --build in the common command usage. What I'm trying to do is to avoid overloading the meaning of --version, as although we're removing it from publish and java, it's still being used for specifying esdl version in xsd and wsdl commands.

mayx replied:
I could change the xsd and wsdl to use --ver_ (which nicely corresponds to the url parameter name to specify service version), and restore --version solely to printout build info. Do you guys like that better?

rpastrana replied:
makes sense, maybe the tool version command remains --version and the interface version gets an explicit name --interfaceversion or some shorter derivation of it?

afishbeck replied:
Agreed that we should use --version the way other commands do (like ecl.exe) rather than --build.

Where we do want interface version specified, the long intuitive name should be something like "--interface-version".. for people who haven't memorized the short easy name, something like "-iv", or if you guys have a better idea for the short name I'm open.

"--" should always be a long form very intuitive and explicit name.  So not --ver_.  Single "-" should be a short name, often an abbreviation or single letter.  But again I think -ver_ is a bit to awkward.

mayx replied:
Sounds reasonable, I'll replace --version with --interface-version and -iv, then replace --build with --verison.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: bdcbb70cd6eb5ae5a88da3b87a0d6e1ac59c3e18
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.468 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 5b09435064d56966483679ed924390217cd4667c
Build: failed
Number of Error(s): 6
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
/mnt/disk1/home/vamosax/smoketest/PR-10316/HPCC-Platform/tools/esdlcmd/esdlcmd_shell.cpp:158:39: error: \xe2\x80\x98ESDLOPT_BUILD\xe2\x80\x99 was not declared in this scope
make[2]: *** [tools/esdlcmd/CMakeFiles/esdl.dir/esdlcmd_shell.cpp.o] Error 1
make[1]: *** [tools/esdlcmd/CMakeFiles/esdl.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

mayx commented:
Changes made as suggested. @rpastrana @afishbeck please review again.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: c427fe775fecac5c55cfd958449c04afd61231c9
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.229 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -79,7 +79,6 @@ class EsdlPublishCmdCommon : public EsdlCmdCommon
                 "   --port <port>                WsESDLConfig service port\n"
                 "   -u, --username <name>        Username for accessing WsESDLConfig service\n"
                 "   -pw, --password <pw>         Password for accessing WsESDLConfig service\n"
-                "   --version <ver>              ESDL service version\n"
```
I think we need to keep this line in the base class usage (and update it to "--interface-version,               ESDL interface version\n") since all extensions of this class support should support --version (now --interface-version)
If publish command won't acknowledge this argument, then that class needs to override this usage.

mayx replied:
Actually the --version (now --interface-version) is not being used by other publish commands. A couple of other commands takes a "definition version/sequence#" as parameter but that is specified on the command line directly, not with --version (now --interface-version) option. I don't think any of the publish commands need or should specify the interface version.

rpastrana replied:
ahh.. ok so then   StringAttr optVersionStr; and double  optVersion; are actually being used as "esdl_definition_version"?? maybe we refactor them to esdl_def_ver?

rpastrana commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -94,8 +93,6 @@ class EsdlPublishCmdCommon : public EsdlCmdCommon
             return true;
         if (iter.matchFlag(optWSProcPort, ESDL_OPTION_SERVICE_PORT) || iter.matchFlag(optWSProcPort, ESDL_OPT_SERVICE_PORT))
             return true;
-        if (iter.matchFlag(optVersionStr, ESDLOPT_VERSION))
```
This if needs to stay, and check for ESDLOPT_INTERFACE_VERSION || ESDLOPT_IV... we might also want to refactor optVersionStr and optVersion to reflect their actual purpose which is to hold the InterfaceVersion value as declared by the user.

--UPDATE-- looking through the code, I just realized that this member var is also used to hold the ESDL definition version for EsdlDeleteESDLDefCmd; EsdlBindMethodCmd. I don't like overloading variables, so we can either declare a new member var for those classes to hold the ESDL definition version, or live with the way things are, but then we can't refactor optVersionStr and optVersion  as suggested above

mayx replied:
Yes this is another thing I was thinking to change, but thought it would tie better with another improvement subtask I may make. But maybe changing the variable name now will make it more clear, let me check...

rpastrana replied:
Also, while we're at it...
It seems the usage at the top level might be off:
esdlcmd_shell.cpp:   
esdl [--version] <command> [<args>]\n\n"
should change to 
  esdl <command> [<args>]\n\n"
and --version should be listed as a command

@JamesDeFabia this change might require a doc change (I think an issue is already open, but wanted to point this out)

mayx replied:
I may also refactor the variable name for the definition sequence. What better describes this variable, "definitionVersion" or "defintionSequence", @rpastrana @afishbeck ?

rpastrana replied:
both are appropriate, sequence might be more accurate, but I think we refer to that number as the definition version in other places. I vote definitionVersion

mayx replied:
I agree with moving --version as a command.

rpastrana reviewed (changes requested):
@mayx  A couple of concerns, I might have clouded the issue with my attempt to explain my concerns... don't hesitate to contact me directly if my explanation doesn't make sense.

mayx commented:
Did the variable name refactoring, also added a separate line for --version usage. @rpastrana @afishbeck please review again.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: c48acde7cc85333a367da2733fde66b56a84581e
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.407 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -615,7 +599,7 @@ class EsdlDeleteESDLDefCmd : public EsdlPublishCmdCommon
                     break;
                 case 1:
-                    optVersionStr.set(arg);
+                    optDefinitionVersionStr.set(arg);
```
It seems inconsistent to me that when we bind a service we use a definition ID of the format "MyDefinition.3", but here we use a separate ESDLService and Version "MyDefinition  3".

I prefer the ID type usage... which is how roxie queries work... "myquery.5".

afishbeck replied:
Especially when it looks like we turn it into an ID before sending it to ESP.

If we make that change we may not need the variable anymore, but I also wonder if the name "PublishedVersion" is more descriptive... meaning the particular version of the ESDLService that was published for "myserv.3"

rpastrana replied:
Agreed, my original implementation introduced this inconsistency. Changing things now will break users, but we might have to tear off the band-aid quickly and move on. 
If we make the change, we could make the fix backward compatible, but I'm not sure if that would be worth the effort.

mayx replied:
Sorry missed this string of comments. I was thinking to deal with this inconsistency in a separate issue, but since we're already here and touched on the code already, I may just do it now.

afishbeck commented on tools/esdlcmd/esdlcmd_common.hpp:
```diff
@@ -62,6 +62,8 @@ typedef IEsdlCommand *(*EsdlCommandFactory)(const char *cmdname);
 
 #define ESDLOPT_VERSION                 "--version"
+#define ESDLOPT_INTERFACE_VERSION       "--interface-version"
+#define ESDLOPT_IV                      "-iv"
```
"-iv" is considered an abbreviated or short form of "--interface-version" so there is a naming convention to just add an "_S".   Making it #define ESDLOPT_INTERFACE_VERSION_S       "-iv"

Like this example:

#define ECLOPT_INPUT "--input"
#define ECLOPT_INPUT_S "-in"

afishbeck reviewed (changes requested):
A few comments

mayx commented:
Changed ESDLOPT_IV to ESDLOPT_INTERFACE_VERSION_S, please review again. There're a few other existing options that don't follow this convention, but I think it would be better to address them in a separate refactoring issue.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 35c18b42c66f29a4e8987c1b306909e9739ade1f
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.402 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

mayx committed:
HPCC-18038 Remove the version option from the publish and java command

- For publish and java commands, remove the version option and don't
  pass in version when loading dependencies
- Don't validate version command option for publish and java
- Remove the --version option from the batch file for ECL IDE
- Remove the --version from ReadMeFirst.txt help file
- Refactor the variable name optVersion to optInterfaceVersion for
  xsd and wsdl commands
- Change defintion version to definition id for commands delete and
  bind-method.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented:
Removed optDefinitionVersion variable, changed definition version to definition id for commands delete and bind-method. @afishbeck @rpastrana please review again.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 7a0d3de6ecc58be6a9b5dc5d3ba257b9460cc75f
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.65 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -863,7 +826,7 @@ class EsdlBindMethodCmd : public EsdlBindServiceCmd
                 "   To bind a ESDL Service, provide the target ESP process name\n"
                 "   (esp which will host the service.) \n"
-                "   It is also necessary to provide the Port on which this service is configured to run (ESP Binding),\n"
+                "   It is also necessary to provide the target ESP binding name,\n"
```
Does just specifying the port not work?  I think using port is much more intuitive than ESP binding name... which is usually a confusing unique id.  If it doesn't work, please open a JIRA to add it (to any command line that currently requires ESP binding name).  If it does work the help should still say so (in fact it should be considered the preferred way.

afishbeck replied:
@mayx @rpastrana Btw, in the future when this is fully dynamic the port will still make sense, but the ESP binding name will no longer make sense.

mayx replied:
For this command (bind-method), it doesn't take port as a parameter. I've created an issue for it: https://track.hpccsystems.com/browse/HPCC-18126

afishbeck reviewed (commented):
@mayx some comments

afishbeck reviewed (approved):
@mayx looks good

rpastrana commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -581,18 +560,17 @@ class EsdlDeleteESDLDefCmd : public EsdlPublishCmdCommon
-                "   ESDLServiceDefinitionName         The name of the ESDL service definition to delete\n"
-                "   ESDLServiceDefinitionVersion      The version of the ESDL service definition to delete\n");
+                "esdl delete <ESDLDefinitionID> [command options]\n\n"
+                "   ESDLDefinitionID      The ESDL definition id <esdldefname>.<esdldefver>\n");
```
@JamesDeFabia note this change. It needs to be documented

rpastrana commented on tools/esdlcmd/esdlcmd_shell.cpp:
```diff
@@ -180,14 +180,15 @@ bool EsdlCMDShell::parseCommandLineOptions(ArgvIterator &iter)
            "   wsdl              Generate WSDL from ESDL definition.\n"
            "   publish           Publish ESDL Definition for ESP use.\n"
            "   list-definitions  List all ESDL definitions.\n"
+           "   get-definition    Get ESDL definition.\n"
```
Is this part of a different change?

mayx replied:
@rpastrana get-definition command is already there, it was only missing from the usage.

rpastrana reviewed (commented):
@mayx looks good. I do have a question about the new command "get-definition"

rpastrana reviewed (approved):
@mayx looks good.
